### ISSUE TITLE
sulu_content_load might lead to errors or unexpected results if both pages have a smart_content

### ISSUE THREAD

**Issue by luca-rath** [sulu/sulu#5822]
| Q | A
| --- | ---
| Sulu Version | 2.2

# Actual Behavior

The query parameters of the current request are taken into account in `sulu_content_load`.

An example would be the `SmartContent ContentType`. It retrieves the current page from `$this->requestStack->getCurrentRequest()->get($pageParameter)`.

Because of this, a page `Foo` with a smart_content (currently on page `2` e.g.) loads another page `Bar` using `sulu_content_load()`. Page `Bar` has also a smart_content on it. But because the url of the original request contains `?page=2`, the smart_content of page `Bar` also tries to load the items of page `2`, but that's not that what a user would expect.

Additionally, if page `Bar` does not have 2 pages, a `PageOutOfBoundsException` will be thrown.

# Expected Behavior

The user would expect page `Foo` to display the smart_content items of page `2` (becase the url says `?page=2`) but he would expect page `Bar` to display the smart_content items of page `1`.

# Steps to Reproduce

~~~xml
<!-- config/templates/pages/default.xml -->

<property name="single_page_selection" type="single_page_selection">
    <meta>
        <title lang="en">Single page selection</title>
    </meta>
</property>

<property name="smart_content" type="smart_content">
    <meta>
        <title lang="en">Smart Content</title>
    </meta>

    <params>
        <param name="provider" value="pages"/>
        <param name="max_per_page" value="1"/>
        <param name="page_parameter" value="p"/>
    </params>
</property>
~~~

~~~twig
{# templates/pages/default.html.twig #}

{{ dump(sulu_content_load(content.single_page_selection)) }}
~~~

# Possible Solutions

~~~php
// ContentTwigExtension.php

$request = $this->requestStack->getCurrentRequest();
$duplicatedRequest = $request->duplicate([], [], null, null, []); // This sets query parameters, request parameters and files to an empty array
$this->requestStack->push($duplicatedRequest);
$resolvedStructure = $this->structureResolver->resolve($contentStructure);
$this->requestStack->pop();

return $resolvedStructure;
~~~

**Comment by luca-rath** [comment]
Somehow related to https://github.com/sulu/sulu/pull/5820 and https://github.com/sulu/sulu/issues/5821

**Comment by alexander-schranz** [comment]
For `sulu_content_load` the idea sounds at current state not bad for me the [requestStack](https://github.com/symfony/symfony/blob/fe4e2956e31074a87555fb14884fdb8a65af15e3/src/Symfony/Component/HttpFoundation/RequestStack.php#L35-L38) itself does not throw a `kernel.request` event which is good in this case, else this could make problems. The [request->duplicate](https://github.com/symfony/symfony/blob/fe4e2956e31074a87555fb14884fdb8a65af15e3/src/Symfony/Component/HttpFoundation/Request.php#L453-L495) also looks fine hopefully symfony will not edit this logics.

To be save. We would need todo something like:

```php
$this->requestStack->push($duplicatedRequest);

try {
    $resolvedStructure = $this->structureResolver->resolve($contentStructure);
} finally {
    $this->requestStack->pop();
}
```

To avoid that the requestStack could go out of sync.

**Comment by niklasnatter** [comment]
Should we close this with #5820 and #5835 being merged? 🙂

**Comment by luca-rath** [comment]
Should we maybe display an error if the user tries to save a page with two smart_contents with pagination enabled and the same page parameters?

**Comment by luca-rath** [comment]
I guess you don't like the idea that's described in `Possible Solutions`?

**Comment by niklasnatter** [comment]
The solution sounds okay for me, but I am a bit afraid about potential side effects of pushing an empty request to the stack and I am not sure if it is still needed. With #5835 being merged, the problem is probably solved for the majority of usecases. 

The case that the solution would solve is resolving the value of a paginated `smart_content` property of another page via the `sulu_content_load` function. I dont think that is a common usecase and therefore I would probably not risk the potential side effects personally. But I am okay with keeping this open too 🙂

**Comment by luca-rath** [comment]
Fixed by #5850
